### PR TITLE
sql: check all timestamps for a valid range

### PR DESCRIPTION
**Pull request by mjibson**
Previously we would allow any timestamp supported by chrono, even if they weren't representable in pgwire/postgres or avro. Instead, wrap all timestamps in a struct that verifies they are valid or returns an error. Valid here means the smallest set of dates that are representable in postgres, avro, and chrono.

This unfortunately hits a huge amount of code because many things require infalliable casts from a rust type into a datum, but all of the datetime logic works on chrono objects.

### Motivation

  * This PR adds a known-desirable feature. #7141

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Timestamp types now have a reduced range now within the postgres spec.

### CODE REVIEW

**Comment by mjibson** [comment]
@guswynn Tagging you here to comment on some of my rust choices in the new `adt/timestamp.rs` file. Dunno if there's a better way to do all this.

File: src/repr/src/adt/timestamp.rs
Comment (guswynn) [inline src/repr/src/adt/timestamp.rs:67]: love this pattern

File: src/repr/src/adt/timestamp.rs
Comment (guswynn) [inline src/repr/src/adt/timestamp.rs:579]: this file all seems reasonable to me...the extension traits are a good pattern, and it seems the general process is that they require the times and dates to act like a `chrono::NaiveDate`, or at least be able to generate one

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (jkosh44) [reply src/adapter/src/catalog/builtin_table_updates.rs:175]: Oh because this is always the current time so until 262143-12-31, this will always fit.

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/testdrive/src/action/kafka/verify.rs
Comment (mjibson) [reply src/testdrive/src/action/kafka/verify.rs:617]: This took me so long to figure out how to do, but I love how it turned out. Makes testdrive avro debugging a breeze.

File: src/repr/src/adt/timestamp.rs
Comment (mjibson) [reply src/repr/src/adt/timestamp.rs:114]: They were moved from expr/scalar/func.rs, I don't think they changed at all during that and I didn't invent them either.

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (mjibson) [reply src/adapter/src/catalog/builtin_table_updates.rs:175]: That's the theory! I thiiiink that if the various expects here to panic then we have bigger problems and probably want to panic? But doing this work did explicitly expose "hey, we have a thing that could go wrong here but no error path, what do?".

File: src/repr/src/adt/timestamp.rs
Comment (mjibson) [reply src/repr/src/adt/timestamp.rs:31]: SGTM, but I'm going to leave that to someone else.

**Comment by mjibson** [comment]
I think this PR is ready, but it has to wait for a wipe release to merge, so sitting on it for now.

**Pull request merged**
